### PR TITLE
Fix/show redundant quality requirement as gray

### PR DESCRIPTION
**Pull request by FeepingCreature**
#### Summary
```SUMMARY: Interface "Gray out redundant tool quality requirements"```

#### Describe the solution
The tool requirement code now looks at `has_one` and greys itself out if another tool provides the requirement.
I also refactored the coloring code in requirements.cpp a bit.

### CODE REVIEW

**Comment by FeepingCreature** [comment]
How do I make the astyle check pass...?

**Comment by jbytheway** [comment]
> How do I make the astyle check pass...?

If you have make and astyle available locally you can `make astyle` in the code directory and it will format everything for you.

**Comment by FeepingCreature** [comment]
Ah! Thanks!

File: src/construction.cpp [hunk 1]
```diff
@@ -421,35 +421,45 @@ void construction_menu()
...
+                            current_line << colorize( result_string, color_title );
                             std::vector<std::string> folded_result_string = foldstring( current_line.str(),
                                     available_window_width );
                             current_buffer.insert( current_buffer.end(), folded_result_string.begin(),
                                                    folded_result_string.end() );
 
                             // display description of the result for multi-stages
                             current_line.str( "" );
-                            current_line << _( "Result: " ) << color_data;
+                            current_line << "Result: ";
```
Comment (BevapDin) [inline src/construction.cpp]: Are there any reasons why you removed the translation of "Result: "?

File: src/construction.cpp [hunk 2]
```diff
@@ -421,35 +421,45 @@ void construction_menu()
...
-                            current_line << "</color>";
                             folded_result_string = foldstring( current_line.str(), available_window_width );
                             current_buffer.insert( current_buffer.end(), folded_result_string.begin(),
                                                    folded_result_string.end() );
 
                             // display description of the result for single stages
                         } else if( !current_con->post_terrain.empty() ) {
                             current_line.str( "" );
-                            current_line << _( "Result: " ) << color_data;
+                            current_line << "Result: ";
```
Comment (BevapDin) [inline src/construction.cpp]: Same here.

File: src/requirements.cpp [hunk 3]
```diff
@@ -432,8 +432,7 @@ std::vector<std::string> requirement_data::get_folded_components_list( int width
     if( components.empty() ) {
         return out_buffer;
     }
-    out_buffer.push_back( "<color_" + string_from_color( col ) + ">" + _( "Components required:" ) +
-                          "</color>" );
+    out_buffer.push_back( colorize( "Components required:", col ) );
```
Comment (BevapDin) [inline src/requirements.cpp]: And here.

File: src/construction.cpp (hunk 1 above)
Comment (FeepingCreature) [reply src/construction.cpp]: Oh, that's what it does! I see. Apologies. I thought it was related to string construction somehow, but now that you mention it `_()` does ring a bell. Oops.

File: src/construction.cpp (hunk 2 above)
Comment (FeepingCreature) [reply src/construction.cpp]: Oh! Apologies.

File: src/requirements.cpp (hunk 3 above)
Comment (FeepingCreature) [reply src/requirements.cpp]: Sorry.

File: src/requirements.cpp [hunk 4]
```diff
@@ -455,32 +467,25 @@ std::vector<std::string> requirement_data::get_folded_list( int width,
...
+                color = yellow_background( color );
             }
-            unfolded += *a;
+
+            list_as_string.push_back( colorize( text, color ) );
+            buffer_has.push_back( text + color_tag );
         }
+        std::sort( list_as_string.begin(), list_as_string.end() );
+
+        const std::string unfolded = join( list_as_string, colorize( " OR ", c_white ) );
```
[the remark this answers is not in the document]
Comment (FeepingCreature) [reply src/requirements.cpp]: Sorry.

File: src/requirements.cpp [hunk 5]
```diff
@@ -498,11 +503,9 @@ std::vector<std::string> requirement_data::get_folded_tools_list( int width, nc_
         const inventory &crafting_inv, int batch ) const
 {
     std::vector<std::string> output_buffer;
-    output_buffer.push_back( "<color_" + string_from_color( col ) + ">" + _( "Tools required:" ) +
-                             "</color>" );
+    output_buffer.push_back( colorize( "Tools required:", col ) );
     if( tools.empty() && qualities.empty() ) {
-        output_buffer.push_back( "<color_" + string_from_color( col ) + ">> </color><color_" +
-                                 string_from_color( c_green ) + ">" + _( "NONE" ) + "</color>" );
+        output_buffer.push_back( colorize( "> ", col ) + colorize( "NONE", c_green ) );
```
[the remark this answers is not in the document]
Comment (FeepingCreature) [reply src/requirements.cpp]: Oops.

File: src/requirements.cpp [hunk 6]
```diff
@@ -498,11 +503,9 @@ std::vector<std::string> requirement_data::get_folded_tools_list( int width, nc_
         const inventory &crafting_inv, int batch ) const
 {
     std::vector<std::string> output_buffer;
-    output_buffer.push_back( "<color_" + string_from_color( col ) + ">" + _( "Tools required:" ) +
-                             "</color>" );
+    output_buffer.push_back( colorize( "Tools required:", col ) );
```
[the remark this answers is not in the document]
Comment (FeepingCreature) [reply src/requirements.cpp]: S.

**Comment by narc0tiq** [comment]
I think I know what this is doing, but is there any chance I could get a screenshot from a recipe where this happens?

**Comment by FeepingCreature** [comment]
Sure, here you go https://i.imgur.com/1iZrtGQ.png

**Comment by FeepingCreature** [comment]
Yes, that's the bugged state. It's gray with the fix.

**Comment by ZhilkinSerg** [comment]
I am not sure you should edit `.po` files directly.

What do you say, @BrettDong?

**Comment by FeepingCreature** [comment]
Oh, is that wrong? What should I edit instead?

**Comment by FeepingCreature** [comment]
Well, I'm not really trying to translate anything new, I just didn't want to do something like `" " + _( "OR" ) + " "` and there was precedent for strings with spaces. Istm there's two relevant questions: are there repositories of strings other than the .po files, and are the .po files authoritive?

**Comment by cainiaowu** [comment]
It is just that anything you change will get reverted in the next po script update. Since all string is extracted by python script from the C code.
You need to find where it is in the C code and change from there.

**Comment by FeepingCreature** [comment]
No, the change is *because* of a change I made in the C++ code. But seen like that, I probably shouldn't have edited it manually at all? I just didn't want people to retranslate things because of a change from `"OR"` to `" OR "`.

**Comment by FeepingCreature** [comment]
:( okay. Will do when I get home.

**Comment by BrettDong** [comment]
Yes, please don't touch anything inside `lang/po`. All files under that folder are automatically generated.

**Comment by FeepingCreature** [comment]
ping

**Pull request merged**